src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.AvoidUnusedMembers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class AvoidUnusedMembersDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        internal static string UnunsedMemberNameProperty = nameof(UnunsedMemberNameProperty);
+        internal static string UnunsedMemberKindProperty = nameof(UnunsedMemberKindProperty);
+
+        private readonly DiagnosticDescriptor _noReadsUnnecessaryWithSuggestionDescriptor;
+
+        public AvoidUnusedMembersDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.AvoidUnusedMembersDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _noReadsUnnecessaryWithSuggestionDescriptor = new DiagnosticDescriptor(
+                UnnecessaryWithSuggestionDescriptor.Id,
+                UnnecessaryWithSuggestionDescriptor.Title,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Members_Without_Reads_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                UnnecessaryWithSuggestionDescriptor.Category,
+                UnnecessaryWithSuggestionDescriptor.DefaultSeverity,
+                UnnecessaryWithSuggestionDescriptor.IsEnabledByDefault,
+                UnnecessaryWithSuggestionDescriptor.Description,
+                UnnecessaryWithSuggestionDescriptor.HelpLinkUri,
+                UnnecessaryWithSuggestionDescriptor.CustomTags.ToArray());
```

Comment:
does DiagnosticDescriptor not have a With'er?

Reply:
Nope.

---
In reply to: [212729411](https://github.com/dotnet/roslyn/pull/29511#discussion_r212729411) [](ancestors = 212729411)

Reply:
Note: instead of manually constructing the descriptor, you can use the helpers defined in the base class.  That would help remove a lot of redundancy here. #unresolved

Reply:
I am going to combine this into a single message/option/ID for now.

---
In reply to: [212774354](https://github.com/dotnet/roslyn/pull/29511#discussion_r212774354) [](ancestors = 212774354)